File: typescript/src/auth/cookie-auth.ts
/**
 * Cookie-based authentication strategy for the Fizzy SDK.
 *
 * CookieAuth sets the Cookie header for session-based authentication,
 * used by mobile and web clients.
 */

import type { AuthStrategy } from "../auth-strategy.js";

/**
 * Provider for cookie values — either a static string or an async function.
 */
export type CookieProvider = string | (() => Promise<string>);

/**
 * Cookie-based authentication strategy.
 * Sets the Cookie header with the provided session cookie.
 */
export class CookieAuth implements AuthStrategy {
  private cookieProvider: CookieProvider;

  constructor(cookieProvider: CookieProvider) {
    this.cookieProvider = cookieProvider;
  }

  async authenticate(headers: Headers): Promise<void> {
    const cookie =
      typeof this.cookieProvider === "function"
        ? await this.cookieProvider()
        : this.cookieProvider;
    headers.set("Cookie", cookie);
  }
}

/**
 * Creates a CookieAuth strategy from a CookieProvider.
 */
export function cookieAuth(cookieProvider: CookieProvider): AuthStrategy {
  return new CookieAuth(cookieProvider);
}

File: swift/Sources/Fizzy/Auth/CookieAuth.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Cookie-based authentication strategy for session-based auth.
///
/// Sets the `Cookie` header with `session_token=<value>` for each request.
/// Used for mobile and web session authentication in Fizzy.
///
/// ```swift
/// let client = FizzyClient(
///     auth: CookieAuth(sessionToken: "your-session-token"),
///     userAgent: "my-app/1.0"
/// )
/// ```
public struct CookieAuth: AuthStrategy {
    private let tokenProvider: any TokenProvider

    /// Creates a CookieAuth strategy with a static session token.
    ///
    /// - Parameter sessionToken: The session token value.
    public init(sessionToken: String) {
        self.tokenProvider = StaticTokenProvider(sessionToken)
    }

    /// Creates a CookieAuth strategy with a custom token provider.
    ///
    /// Use this for dynamic session token scenarios (e.g., rotation).
    ///
    /// - Parameter tokenProvider: A provider that returns session tokens.
    public init(tokenProvider: any TokenProvider) {
        self.tokenProvider = tokenProvider
    }

    public func authenticate(_ request: inout URLRequest) async throws {
        let token = try await tokenProvider.accessToken()
        request.setValue("session_token=\(token)", forHTTPHeaderField: "Cookie")
    }
}
